GCloud backend

pbzdyl opened the pull request:
Hello,

I have implemented a new backend: GCloud. It uses Google Storage for persisting state files and for distributed locks. It supports multiple workspaces.

Thanks to the authors of remote-state/consul and remote-state/s3 as I used their implementation as a basis for my code. Disclaimer: I am new to Go and Terraform, so please forgive me any dumbness in the code!

Features:
- remote locking
- multiple workspaces
- uses account key JSON file (`credentials` option, either file path to the JSON file or directly JSON contents)
- if `credentials` is not specified, uses [Google Application Default Credentials](https://developers.google.com/identity/protocols/application-default-credentials#howtheywork)

Example configuration:

```
terraform {
  backend "gcloud" {
    bucket = "my-bucket"
    state_dir = "terraform-state"
    credentials = "path/to/account-key.json"
  }
}
```

Actual state file URL will be: `gs://${bucket}/${state_dir}/${workspace_name}.tfstate`
and a lock file will be created at: `gc://${bucket}/${state_dir}/$workspace_name}.tflock`

State locks use [Google Cloud Storage Concurrency Control mechanism](https://cloud.google.com/storage/docs/gsutil/addlhelp/ObjectVersioningandConcurrencyControl#concurrency-control).

I am aware there is already the `gcs` backend. However, my understanding is that it is a legacy remote client implementation which doesn't support multiple workspaces. There would also be a risk of regressions potentially impacting deployments that use it.

This is of course still work in progress as I am not sure if this PR would be accepted or not. If there is a chance it could be merged, I would complete the PR with tests, documentation updates and anything suggested by the reviewers.

Best regards,
Piotr

geerew commented:
👍

sl1pm4t commented:
Hi @pbzdyl 
I'm glad my colleague pointed this PR out to me, because I was about to start work on implementing a GCS backend.

I contributed the legacy `gcs` remote last year, and you're correct that it doesn't support workspaces or locking etc - simply because those concepts didn't exist in Terraform back then.

If this backend is accepted, I don't see any need to keep the existing `gcs` remote. Just like the legacy `s3` remote was removed when the `s3` backend was implemented.

I'll give this backend a try - thanks!

nodesocket commented:
👍 for this replacing the existing `gcs` integration if everything checks out.

pbzdyl commented:
Hello @sl1pm4t ,

Thanks! Please, do share your feedback from testing this PR. If you have any suggestions for the PR's code, I would be happy to hear too as I am new to terraform and Go so I am sure there is something I could improve.

I am not sure removing the current `gcs` implementation and replacing it with this one using the same name would be welcomed by users of the current `gcs` as it would be backward incompatible. By using a different name, users could migrate from one backend to another by changing the configuration and moving their state.

octo commented:
Hi @pbzdyl, thank you very much for your work!

I was about to start working on implementing locking for the GCS backend when I stumbled into this PR. I'm quite interested in moving this along; are you still interested in working on this? I'd be happy to collaborate on this; I could e.g. send PRs against your feature branch if that's okay with you …?

Best regards,
—octo

pbzdyl commented:
Hello @octo,

Yes - definitely. However, I am not sure if this PR will be accepted as there has been no feedback from Terraform team.

What kind of changes would you like to add to this PR?

Regards,
Piotr

octo commented:
@pbzdyl Great! Mostly coding style changes for now, but there are two noteworthy changes:

1.  Return the GCS object's *generation* as the lock ID. This allows `Unlock()` to call "delete if generation matches" immediately, removing one round-trip.
2.  Hold on to the context passed to the `configure()` callback to avoid calling `context.Background()`.

The changes are in the [pr/15592](https://github.com/octo/terraform/tree/pr/15592) branch. I'll open a PR against your repo in a minute.

Thanks and best regards,
—octo

jbardin commented:
Hi,

Sorry about the delayed response here, and thanks for the work you put into this!
I hope to catch up with a review shortly, but I have a preliminary question; Are there some incompatibilities that are preventing updating the existing "gcs" remote state to a backend? We generally would prefer a transparent upgrade path for existing users, rather than creating an entirely new backend.

pbzdyl commented:
Hello @jbardin,

Do you mean removing current `gcs` implementation and renaming the new backend to `gcs`? I am not sure I have enough knowledge about the internals but I will try to implement it this way and test it locally.

jbardin commented:
Hi @octo,

I think the existing "remote-state" backends were all migrated from legacy remote state implementations, with azure being the most recent. S3 might be a good learning example, as it probably has the most "hacks" for legacy behavior, and is one of the more complex requiring S3 and DynamoDB in concert.

octo commented:
Great, thanks @jbardin!

octo commented:
Hi everybody!

pbzdyl/terraform#2 has a lot of updates to this PR and should appear here once merged by @pbzdyl.

How do you want to deal with the merge conflicts in vendor/? Should I rebase onto or merge the master branch? I can also remove the commit that changes vendor/ and we can do this as the last commit of the PR.

Best regards,
—octo

octo commented:
Thank you @pbzdyl!

@jbardin I've done all the changes I was intending to make; from my POV, this is good to go. If anything's missing, please let me know and I'll make sure to add / change that.

Best regards,
—octo

jbardin commented:
Hi @octo,

Thanks for the update, I'll take a look ASAP. 
As for the conflicts, we usually prefer to rebase on master then force push the new commits into the PR. The vendor files however are a little different, since you can't "resolve" conflicts in there. I would drop the first commit, rebase on master, then use govendor to update the missing dependencies. 

Don't worry if you can't get the vendor files to work out. If it turns into too much of a mess, I can transfer this whole PR into a branch here and fix them at the end.

pbzdyl commented:
I have removed the commit updating vendor files and rebased on the latest master branch to resolve merge conflicts. I will try to update vendor files tomorrow.

pbzdyl commented:
@jbardin, I was able to update vendor files.

pbzdyl commented:
I was testing the new backend today and I noticed that we probably should add an ignored `project` configuration property to make the new backend fully compatible with the old one so that users with existing gcs configuration can start using the new one without any configuration changes.

octo commented:
I just tried the new code for the first time, and there are some small issues that need fixing:

*   The old code had a `project` option which the new code is missing. I'll have to find out what it was used for and do something sensible.
*   The `credentials` option is evaluated relative to the process's working directory, not the *terraform working directory* which is unexpected. I.e. if `example.tf` references `creds.json`, this should be interpreted as relative to `example.tf`, not the (meaningless) directory from which `terraform init /path/to/terraform/wd` was executed.
*   We should make an effort to create the bucket for the user if it doesn't exist. This is a UX issue that only pops up when initializing, but I think is worth fixing in code to give new users an easier time getting started.

I'll send patches for those issues soon. If you know a way to determine the terraform working directory (without having access to `command.Meta.DataDir()`, I assume) I'd love to hear it :)

Best regards,
—octo

octo commented:
@pbzdyl Thanks for pointing out the `Schema.Deprecated` field, I was not aware of the previously. I'd say that the `path` option should definitely have that.

I just researched the `project` option, but it appears that it's only present in the config but not actually used by the old code. We *could* add a new *Removed* option so users who copied this from the docs know that it doesn't have any effect, but IMHO that would break backwards compatibility. @jbardin, what do you think?

—octo

pbzdyl commented:
@octo I have added you as a collaborator to my fork so you can push directly to this PR's branch without hassle of using PRs.

octo commented:
Awesome, thanks @pbzdyl!

octo commented:
@pbzdyl Yes, strictly speaking that breaks backwards compatibility: previously, `pathorcontents` allowed you to do is store the JSON credentials *inline*, i.e. have a string to the Terraform config that contained (escaped) JSON. The downside is that if the file isn't found, e.g. because the relative path lookup works differently than the user expects, then you get a *parse error* rather than a *file not found error*, which is not a great user experience.

If that feature is actually used, we should probably revert that change. If not, I'd vote for removing unnecessary complexity. Any thoughts?

Best regards,
—octo

octo commented:
@pbzdyl Hi Piotrek,

could you delete the "[WIP]" prefix from this PR, as per the [contributing guidelines](https://github.com/hashicorp/terraform/blob/master/.github/CONTRIBUTING.md)? Thank you!

—octo

octo commented:
@jbardin Ping?

jbardin commented:
Hi @octo,

(Sorry about the delay here, last week was quite busy to say the least.)

This is looking good, but I would like to make sure the ACC tests can run (I haven't been able to due to always seeing duplicate buckets for some reason). Due to the fact that ACC tests are still a manual process for backends, we usually paste the passing test output into a comment here. 

We also needs to add tests that call `remote.TestRemoteLocks`, as well as `backend.TestBackend` with multiple backend instances. 

I would like to merge this soon, but also want to wait until we have a handle on "known urgent" issues, so that we can have some more time to test this out in main before releasing.

Thanks!

octo commented:
Thanks for your feedback @jbardin!

The duplicate buckets may be due to 5f4061440faa6ced2c265c835175567e045681a0, which may conflict with existing tests. I'll make sure that works and come back to you.

Tests based on `remote.TestRemoteLocks` and `backend.TestBackend` are coming right up.

Best regards,
—octo

octo commented:
```
octo@didactylos:~/go/src/github.com/hashicorp/terraform/backend/remote-state/gcs$ git log --pretty=oneline HEAD^..HEAD; go test
5155a55ef63ca13b8bce46d8f8dbb1edddfbaa57 (HEAD -> gcloud-backend) backend/remote-state/gcs: Implement additional tests.
2017/09/26 11:26:30 [TRACE] Preserving existing state lineage "7f564c98-cb5c-4338-848a-a0423e60029e"
2017/09/26 11:26:30 [TRACE] Preserving existing state lineage "7f564c98-cb5c-4338-848a-a0423e60029e"
2017/09/26 11:26:30 [DEBUG] New state was assigned lineage "74616807-1ec5-41f7-b753-0c4b5eb4fc31"
2017/09/26 11:26:30 [TRACE] Preserving existing state lineage "74616807-1ec5-41f7-b753-0c4b5eb4fc31"
2017/09/26 11:26:31 [DEBUG] New state was assigned lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:31 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:31 [DEBUG] New state was assigned lineage "508cbca7-2094-4564-b17e-eb19975abfe9"
2017/09/26 11:26:31 [TRACE] Preserving existing state lineage "508cbca7-2094-4564-b17e-eb19975abfe9"
2017/09/26 11:26:31 [TRACE] Preserving existing state lineage "7f564c98-cb5c-4338-848a-a0423e60029e"
2017/09/26 11:26:31 [TRACE] Preserving existing state lineage "7f564c98-cb5c-4338-848a-a0423e60029e"
2017/09/26 11:26:32 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:32 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:33 [DEBUG] New state was assigned lineage "55ab1489-5ed6-4b03-b2e0-5c2f2b2c5b63"
2017/09/26 11:26:33 [TRACE] Preserving existing state lineage "55ab1489-5ed6-4b03-b2e0-5c2f2b2c5b63"
2017/09/26 11:26:34 [TRACE] Preserving existing state lineage "55ab1489-5ed6-4b03-b2e0-5c2f2b2c5b63"
2017/09/26 11:26:34 [TRACE] Preserving existing state lineage "55ab1489-5ed6-4b03-b2e0-5c2f2b2c5b63"
2017/09/26 11:26:34 [DEBUG] New state was assigned lineage "99171396-9ee5-427c-91b6-45c378cfac90"
2017/09/26 11:26:34 [DEBUG] New state was assigned lineage "06a36fd4-d403-47ea-a2ad-8fd3b19ce550"
2017/09/26 11:26:34 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:34 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:34 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:35 [TRACE] Preserving existing state lineage "55ab1489-5ed6-4b03-b2e0-5c2f2b2c5b63"
2017/09/26 11:26:35 [TRACE] Preserving existing state lineage "55ab1489-5ed6-4b03-b2e0-5c2f2b2c5b63"
2017/09/26 11:26:35 [TRACE] Preserving existing state lineage "55ab1489-5ed6-4b03-b2e0-5c2f2b2c5b63"
2017/09/26 11:26:36 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:36 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:36 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:36 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:37 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:37 [TRACE] Preserving existing state lineage "92bdae17-da0b-4b87-a658-708450136b1e"
2017/09/26 11:26:38 [TRACE] Preserving existing state lineage "55ab1489-5ed6-4b03-b2e0-5c2f2b2c5b63"
2017/09/26 11:26:38 [TRACE] Preserving existing state lineage "55ab1489-5ed6-4b03-b2e0-5c2f2b2c5b63"
2017/09/26 11:26:39 [TRACE] Preserving existing state lineage "55ab1489-5ed6-4b03-b2e0-5c2f2b2c5b63"
2017/09/26 11:26:39 [TRACE] Preserving existing state lineage "55ab1489-5ed6-4b03-b2e0-5c2f2b2c5b63"
2017/09/26 11:26:40 [DEBUG] New state was assigned lineage "09a187b2-11fa-49c4-8a99-c3ed12074d9b"
2017/09/26 11:26:40 [TRACE] Preserving existing state lineage "09a187b2-11fa-49c4-8a99-c3ed12074d9b"
2017/09/26 11:26:41 [TRACE] Preserving existing state lineage "09a187b2-11fa-49c4-8a99-c3ed12074d9b"
2017/09/26 11:26:41 [TRACE] Preserving existing state lineage "09a187b2-11fa-49c4-8a99-c3ed12074d9b"
2017/09/26 11:26:42 [TRACE] Preserving existing state lineage "fe3bffde-285e-4de2-86d1-cbc7f622a277"
2017/09/26 11:26:42 [TRACE] Preserving existing state lineage "fe3bffde-285e-4de2-86d1-cbc7f622a277"
2017/09/26 11:26:43 [TRACE] Preserving existing state lineage "fe3bffde-285e-4de2-86d1-cbc7f622a277"
2017/09/26 11:26:43 [TRACE] Preserving existing state lineage "fe3bffde-285e-4de2-86d1-cbc7f622a277"
2017/09/26 11:26:44 [TRACE] Preserving existing state lineage "fe3bffde-285e-4de2-86d1-cbc7f622a277"
2017/09/26 11:26:44 [TRACE] Preserving existing state lineage "fe3bffde-285e-4de2-86d1-cbc7f622a277"
2017/09/26 11:26:44 [TRACE] Preserving existing state lineage "fe3bffde-285e-4de2-86d1-cbc7f622a277"
2017/09/26 11:26:44 [TRACE] Preserving existing state lineage "fe3bffde-285e-4de2-86d1-cbc7f622a277"
PASS
ok  	github.com/hashicorp/terraform/backend/remote-state/gcs	17.586s
```

kshep commented:
Really looking forward to these features!

I haven't dug through the PR, but will this include instructions for migrating from the existing gcs backend without rebuilding all your infrastructure? Or in the end is it a direct drop in replacement?

octo commented:
@kshep It *should* be a drop-in replacement, but I haven't tested that (yet). Since I don't have an existing gcs-based setup I can't test this with a real world scenario, so help is welcome ;-)

octo commented on backend/remote-state/gcs/backend_test.go:
```diff
@@ -0,0 +1,157 @@
+
+	remote.TestClient(t, rs.Client)
+
+	cleanBackend(t, be)
```
Done

octo commented on backend/remote-state/gcs/backend_test.go:
```diff
@@ -0,0 +1,157 @@
+
+	// clean up all states left behind by previous runs --
+	// backend.TestBackend() will complain about any non-default states.
+	cleanBackend(t, be0)
```
Done

octo commented on backend/remote-state/gcs/backend_test.go:
```diff
@@ -0,0 +1,157 @@
+
+// cleanBackend deletes all states from be except the default state.
+func cleanBackend(t *testing.T, be backend.Backend) {
+	t.Helper()
```
Done

octo commented:
Hi @jbardin, thanks for your comments.

I've fixed them: cleanup is now deferred right after the call to `setupBackend()` and removes the default state and the bucket itself, too.

I also had an epiphany: bucket names need to be *globally* unique, so using essentially hard-coded bucket names (derived from the test names) is not going to work. I've changed the bucket names to include the GCP *project ID* and added sanitation code because project IDs may contain characters that are illegal in bucket names.

Tests are still executed in parallel as I don't see any benefit in running them sequentially.

Best regards,
—octo

octo commented:
*   `toBucketName()` was introduced to allow the tests to generate unique bucket names, I've therefore moved that function to the tests. I did not add a verification function for bucket names, because there's server side verification anyway and worst case we'd refuse a valid bucket name.
*   I updated the `cloud.google.com/go/storage` package in the `vendor/` folder.
*   I rebased onto *master* and resolved the merge conflicts.

The build failure is due to our updating of the vendored gRPC library, which pulls in grpc/grpc-go@596a6acc87c73c3bc398a60ef0089976491ca060. I'll send a fix for that in a moment.

jbardin commented:
@octo: Everything looks good to me, thanks for hanging in there!

I'm going to hold off merging briefly, just until I can get some hands on testing with existing gcs remote states to try and catch any regressions. I do want to get that done ASAP though, so we don't start to lag behind and run into more merge conflicts, and can get this in before we start the process for the next major release.

octo commented:
@jbardin Awesome, thank you! :)

octo commented:
Hi @jbardin, is there any update? Is there anything I can help with to move this forward?

Best regards,
—octo

jbardin commented:
@octo: Thanks, just trying to schedule time to QA a build with this. It needs to be in before the beta process for 0.11 ;)

jbardin commented:
moved this over to a branch where I could resolve the merge conflict asap.

octo commented:
:tada: Thanks @pbzdyl and @jbardin!

pbzdyl commented:
Thank you @octo - all the hard work was done by you!

sl1pm4t commented:
Thanks @pbzdyl + @octo + @jbardin and all who helped with this one!

kavehmzta commented:
All,

Can someone check the following bug that seems to be related to be this PR? 

https://github.com/hashicorp/terraform/issues/16741
